src/io/mod.rs

```diff
@@ -1,5 +1,113 @@
 //! Input and output of images.
+
+use crate::{error, ImageError, ImageResult};
+
 mod reader;
 pub(crate) mod free_functions;
 
 pub use self::reader::Reader;
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+/// Set of supported strict limits for a decoder
+pub struct LimitSupport {
+    /// This field indicates whether the strict limit on the number of bytes is
+    /// supported.
+    pub max_alloc_strict: bool,
+    _non_exhaustive: (),
+}
+
+impl Default for LimitSupport {
+    fn default() -> LimitSupport {
+        LimitSupport {
+            max_alloc_strict: false,
+            _non_exhaustive: (),
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
+/// Resource limits for decoding.
+///
+/// Limits can be either *strict* or *non-strict*. Non-strict limits are best-effort
+/// limits where the library does not guarantee that limit wont be exceeded. Do note 
+/// that it is still considered a bug if a non-strict limit is exceeded, however as 
+/// some of the underlying decoders do . For strict
+/// limits the library makes a stronger guarantee that the limit will not be exceeded.
+/// Exceeding a strict limit is considered a critical bug. If a decoder cannot 
+/// guaranree that it will uphold a strict limit it *must* fail with 
+/// `image::error::LimitErrorKind::Unsupported`.
+///
+/// The limit check should only ever fail if a limit will be exceeded or an unsupported strict
+/// limit is used.
+pub struct Limits {
+    /// The maximum allowed image width. This limit is strict. The default is no limit.
+    pub max_image_width: Option<u32>,
```

Comment:
It would save a some space to make these `Option<NonZeroU32>`s. Though it also would be slightly less ergonomic, so I could be persuaded either way

Reply:
Since this is part of the user facing api I don't think the extra 4 bytes saved is worth it.

Reply:
Well, 16 bytes if you include the other fields which also have zero as a nonsensical limit.